src/EventSubscriber/SearchNoHitEventSubscriber.php

```diff
@@ -57,17 +64,101 @@ public static function getSubscribedEvents(): array
      * Event handler.
      *
      * @param SearchNoHitEvent $event
+     *   Search no hit event
      */
     public function onSearchNoHitEvent(SearchNoHitEvent $event): void
     {
-        if ($this->enabled) {
+        if ($this->noHitsProcessingEnabled) {
+            $keyedNoHits = [];
+
+            /** @var NoHitItem $noHit */
             foreach ($event->getNoHits() as $noHit) {
-                $message = new ProcessMessage();
-                $message->setIdentifierType($noHit->getIsType())
-                    ->setIdentifier($noHit->getIsIdentifier());
+                $cacheKey = $this->getValidCacheKey($noHit->getIsType(), $noHit->getIsIdentifier());
+                $keyedNoHits[$cacheKey] = $noHit;
+            }
+
+            $nonCommittedCacheItems = $this->getNonCachedNoHits($keyedNoHits);
+            $this->sendSearchNoHitEvents($nonCommittedCacheItems);
+        }
+    }
+
+    /**
+     * Send search no hit events.
+     *
+     * @param array $nonCommittedCacheItems
+     *   Array of cache items
+     */
+    private function sendSearchNoHitEvents(array $nonCommittedCacheItems): void
+    {
+        foreach ($nonCommittedCacheItems as $cacheItem) {
+            /** @var NoHitItem $noHitItem */
+            $noHitItem = $cacheItem->get();
+            $message = new ProcessMessage();
+            $message->setIdentifierType($noHitItem->getIsType())
+                ->setIdentifier($noHitItem->getIsIdentifier());
+
+            $this->noHitsCache->saveDeferred($cacheItem);
+
+            $this->producer->sendEvent('SearchNoHitsTopic', JSON::encode($message));
+        }
 
-                $this->producer->sendEvent('SearchNoHitsTopic', JSON::encode($message));
+        $this->noHitsCache->commit();
+    }
+
+    /**
+     * Get cache items for the identifiers not present in the cache.
+     *
+     * @param array $keyedNoHits
+     *   Array og cacheKey => NoHitItem pairs
+     *
+     * @return array
+     *   Array of cache items not yet committed to cache
+     */
+    private function getNonCachedNoHits(array $keyedNoHits): array
+    {
+        $nonCommittedCacheItems = [];
+        try {
+            $cacheKeys = array_keys($keyedNoHits);
+            $cacheItems = $this->noHitsCache->getItems($cacheKeys);
+            foreach ($cacheItems as $cacheItem) {
+                if (!$cacheItem->isHit()) {
+                    /** @var NoHitItem $noHitItem */
+                    $cacheKey = $cacheItem->getKey();
+                    $noHitItem = $keyedNoHits[$cacheKey];
+                    $cacheItem->set($noHitItem);
```

Comment:
So this `getNonCachedNoHits` function sets entries into cache?

Reply:
Which would make the `$nonCommittedCacheItems` more like `$committedCacheItems`?

Reply:
Should these items not have a configureable expire set?

Reply:
> So this `getNonCachedNoHits` function sets entries into cache?

No. It sets the value to be cached. But it does not commit to the cache.

Reply:
> Which would make the `$nonCommittedCacheItems` more like `$committedCacheItems`?

They are not commited to cache until a `ProcessMessage` has been sent in `sendSearchNoHitEvents()`

Reply:
> Should these items not have a configureable expire set?

They are cached in a dedicated cache pool, configured with `default_lifetime: 172800`

Reply:
I can see a point in refactoring `getNonCachedNoHits()` and `sendSearchNoHitEvents()` so that all cache interactions happen in `getNonCachedNoHits()` (and then rename that method).

Should I do that?

Reply:
Naaa... now that I see where you save the cache entry it make more sens :-)